### PR TITLE
feat: PPT-1007 Add DER codec and refactor by extending modules with self

### PR DESCRIPTION
**Pull request by grkek**
- Added the DER codec
- Modified other codecs to extend self for ease of use

### CODE REVIEW

**Comment by stakach** [comment]
You could support 64 bit integers using this algorithm from my implementation of this:
https://github.com/spider-gazelle/bindata/blob/master/src/bindata/asn1/data_types.cr

although not sure if required?

**Comment by grkek** [comment]
> https://github.com/spider-gazelle/bindata/blob/master/src/bindata/asn1/data_types.cr

Not required, removed because of that

File: spec/matter_spec.cr [hunk 1]
```diff
@@ -58,5 +57,26 @@ describe Matter do
...
+      "signAlgorithm" => Matter::Codec::DERCodec::EcdsaWithSHA256_X962.call,
+      "signature" => Matter::Codec::DERCodec::ByteArray.new(Slice(UInt8).new(signature.size) { |i| signature[i] }).value,
+    } of String => Matter::Codec::DERCodec::Value
+
+    encoded = "30 81 cb 30 71 02 01 00 30 0e 31 0c 30 0a 06 03 55 04 0a 0c 03 43 53 52 30 5a 30 13 06 07 2a 86 48 ce 3d 02 01 06 08 2a 86 48 ce 3d 03 01 07 03 43 00 00 04 4b d6 87 ab d2 9b 59 d8 b1 2e 8c 66 14 bd 16 64 ad b2 d4 02 45 5b 6c a3 ef 4e 58 1e 3b e3 44 b8 32 12 e6 14 f2 7e a4 ee c8 f3 1c 75 74 74 38 73 9b 1d 45 1a 7e ab 3a 30 54 2a 0a 7d 18 82 a4 59 a0 00 30 0a 06 08 2a 86 48 ce 3d 04 03 02 03 4a 00 00 30 46 02 21 00 80 86 1a d5 36 ef f0 1c ad 42 81 6a 81 72 f7 1b e3 e4 fd 72 30 cf 73 a4 5e 34 94 5f e8 9d 5d 72 02 21 00 87 fc 1f 47 ad b6 d1 50 58 07 06 86 5e 2e 21 e2 96 3c 9c 15 00 6b 64 da b5 65 8b fb 98 0a 2a d3".split.map(&.to_u8(16))
+
+    crafted = Matter::Codec::DERCodec::Base.encode(packet)
+    original = Slice(UInt8).new(encoded.size) { |i| encoded[i] }
+
+    crafted.hexstring.should eq original.hexstring
```
Comment (stakach) [inline spec/matter_spec.cr:79]:
```crystal
encoded = "30 81 cb 30 71 ...".gsub(' ', "").hexbytes
crafted = Matter::Codec::DERCodec::Base.encode(packet)
crafted.hexstring.should eq encoded.hexstring
```

removes the need for the `original = Slice(UInt8).new(encoded.size) { |i| encoded[i] }` line

**Review by stakach — CHANGES_REQUESTED** [review]

File: spec/matter_spec.cr (hunk 1 above)
Comment (grkek) [reply spec/matter_spec.cr:79]: Oh my, what a wonderful way to do that.

**Pull request merged**